File: services/web-shell/src/telemetry.ts
import { logs, SeverityNumber } from '@opentelemetry/api-logs';

const OTEL_ENDPOINT_DEFAULT = 'http://otel-collector.rube-goldberg:4317';

type NodeProc = {
  versions: { node: string };
  on: (event: string, listener: (...args: unknown[]) => void) => void;
  exit: (code?: number) => void;
  env: Record<string, string | undefined>;
};

function nodeProcess(): NodeProc | undefined {
  return (globalThis as unknown as { process?: NodeProc }).process;
}

function isNodeRuntime(): boolean {
  const proc = nodeProcess();
  return proc != null && typeof proc.versions?.node === 'string' && typeof proc.on === 'function';
}

function otel(name: string): string {
  return `@opentelemetry/${name}`;
}

let initialized = false;

export async function initTelemetry(
  serviceName: string,
  serviceVersion?: string,
): Promise<boolean> {
  if (initialized) return false;
  initialized = true;

  if (!isNodeRuntime()) {
    console.log(
      `[${serviceName}] ${serviceVersion ? `${serviceVersion} ` : ''}startup: telemetry disabled (non-Node runtime)`,
    );
    return false;
  }

  const proc = nodeProcess()!;

  try {
    const endpoint = (proc.env.OTEL_EXPORTER_OTLP_ENDPOINT || OTEL_ENDPOINT_DEFAULT).trim();

    const { NodeSDK } = (await import(otel('sdk-node'))) as typeof import('@opentelemetry/sdk-node');
    const { SimpleLogRecordProcessor } = (await import(otel('sdk-logs'))) as typeof import('@opentelemetry/sdk-logs');
    const { OTLPTraceExporter } = (await import(otel('exporter-trace-otlp-grpc'))) as typeof import('@opentelemetry/exporter-trace-otlp-grpc');
    const { OTLPLogExporter } = (await import(otel('exporter-logs-otlp-grpc'))) as typeof import('@opentelemetry/exporter-logs-otlp-grpc');
    const { getNodeAutoInstrumentations } = (await import(otel('auto-instrumentations-node'))) as typeof import('@opentelemetry/auto-instrumentations-node');

    const sdk = new NodeSDK({
      serviceName,
      traceExporter: new OTLPTraceExporter({ url: endpoint }),
      logRecordProcessors: [
        new SimpleLogRecordProcessor({ exporter: new OTLPLogExporter({ url: endpoint }) }),
      ],
      instrumentations: [getNodeAutoInstrumentations()],
    });

    sdk.start();

    logs.getLogger('rghw').emit({
      severityNumber: SeverityNumber.INFO,
      body: 'service started',
      attributes: {
        'service.name': serviceName,
        ...(serviceVersion ? { 'service.version': serviceVersion } : {}),
      },
    });

    const shutdown = async () => {
      try {
        await sdk.shutdown();
      } catch {
        // best-effort flush of telemetry on exit
      }
      proc.exit(0);
    };
    proc.on('SIGTERM', shutdown);
    proc.on('SIGINT', shutdown);

    return true;
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err);
    console.warn(`[${serviceName}] telemetry unavailable, running without OTLP export: ${detail}`);
    return false;
  }
}

File: services/vector-normalizer-go/internal/telemetry/telemetry.go
// Package telemetry exports vector-normalizer traces and logs to an
// OpenTelemetry collector over OTLP/gRPC.
package telemetry

import (
	"context"
	"strings"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/log"
	"go.opentelemetry.io/otel/log/global"
	"go.opentelemetry.io/otel/propagation"
	sdklog "go.opentelemetry.io/otel/sdk/log"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"

	"rghw.dev/vector-normalizer/internal/version"
)

// ServiceName is the resource service.name reported for every signal.
const ServiceName = "vector-normalizer"

// DefaultEndpoint is the in-cluster collector address used when
// OTEL_EXPORTER_OTLP_ENDPOINT is unset.
const DefaultEndpoint = "http://otel-collector.rube-goldberg:4317"

const scopeName = "rghw.dev/vector-normalizer"

// shutdownTimeout bounds the final flush so an unreachable collector cannot
// stall service exit; tests override it.
var shutdownTimeout = 2 * time.Second

// Init installs global tracer and logger providers exporting over OTLP/gRPC
// to endpoint, or to DefaultEndpoint when endpoint is empty. Export is best
// effort: a signal whose exporter cannot be created is left disabled and the
// service runs on. The returned function flushes and stops both providers and
// is always safe to call.
func Init(ctx context.Context, endpoint string) func() {
	target := grpcTarget(endpoint)
	res := resource.NewWithAttributes(semconv.SchemaURL,
		semconv.ServiceName(ServiceName),
		semconv.ServiceVersion(version.Version),
	)
	stops := make([]func(context.Context) error, 0, 2)

	if exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(target),
		otlptracegrpc.WithInsecure(),
	); err == nil {
		provider := sdktrace.NewTracerProvider(
			sdktrace.WithResource(res),
			sdktrace.WithBatcher(exporter),
		)
		otel.SetTracerProvider(provider)
		stops = append(stops, provider.Shutdown)
	}

	if exporter, err := otlploggrpc.New(ctx,
		otlploggrpc.WithEndpoint(target),
		otlploggrpc.WithInsecure(),
	); err == nil {
		provider := sdklog.NewLoggerProvider(
			sdklog.WithResource(res),
			sdklog.WithProcessor(sdklog.NewBatchProcessor(exporter)),
		)
		global.SetLoggerProvider(provider)
		stops = append(stops, provider.Shutdown)
	}

	otel.SetTextMapPropagator(propagation.TraceContext{})

	return func() {
		ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
		defer cancel()
		for _, stop := range stops {
			_ = stop(ctx)
		}
	}
}

// Info emits an informational log record. Bodies and attributes must describe
// only control flow: never the requested plaintext, expected characters,
// glyph geometry, or image bytes.
func Info(ctx context.Context, body string, attrs ...log.KeyValue) {
	emit(ctx, log.SeverityInfo, "INFO", body, attrs...)
}

// Error emits an error log record under the same payload restrictions as Info.
func Error(ctx context.Context, body string, attrs ...log.KeyValue) {
	emit(ctx, log.SeverityError, "ERROR", body, attrs...)
}

func emit(ctx context.Context, severity log.Severity, severityText, body string, attrs ...log.KeyValue) {
	var record log.Record
	record.SetTimestamp(time.Now())
	record.SetSeverity(severity)
	record.SetSeverityText(severityText)
	record.SetBody(log.StringValue(body))
	record.AddAttributes(attrs...)
	global.GetLoggerProvider().Logger(scopeName).Emit(ctx, record)
}

func grpcTarget(endpoint string) string {
	target := strings.TrimSpace(endpoint)
	if target == "" {
		target = DefaultEndpoint
	}
	if i := strings.Index(target, "://"); i >= 0 {
		target = target[i+3:]
	}
	return strings.TrimSuffix(target, "/")
}
